src/core/expr/head_func_cut.cc

```diff
@@ -0,0 +1,214 @@
+                               break;
+      case dt::SType::FLOAT32: coli = Column(Cut_ColumnImpl::make<double>(std::move(coli), nbins[i], right_closed_));
+                               break;
+      case dt::SType::FLOAT64: coli = Column(Cut_ColumnImpl::make<double>(std::move(coli), nbins[i], right_closed_));
+                               break;
+      default:  throw TypeError() << "cut() can only be applied to numeric "
+                  << "columns, instead column `" << i << "` has an stype: `"
+                  << coli.stype() << "`";
+    }
+    wf.replace_column(i, std::move(coli));
+  }
+
+  return wf;
+}
+
+}}  // dt::expr
+
+
+namespace py {
+
+static oobj make_pyexpr(dt::expr::Op opcode, otuple targs, otuple tparams) {
+  size_t op = static_cast<size_t>(opcode);
+  return robj(Expr_Type).call({ oint(op), targs, tparams });
+}
+
+
+static oobj cut_frame(oobj arg0, oobj arg1, oobj arg2) {
+  auto slice_all = oslice(oslice::NA, oslice::NA, oslice::NA);
+  auto f_all = make_pyexpr(dt::expr::Op::COL,
+                           otuple{ slice_all },
+                           otuple{ oint(0) });
+  auto cutexpr = make_pyexpr(dt::expr::Op::CUT,
+                               otuple{ f_all },
+                               otuple{ arg1, arg2 });
+  auto frame = static_cast<Frame*>(arg0.to_borrowed_ref());
+  return frame->m__getitem__(otuple{ slice_all, cutexpr });
+}
+
+
+
+static PKArgs args_cut(
+  1, 2, 0, false, false,
+  {
+    "cols", "nbins", "right_closed"
+  },
+  "cut",
+
+R"(cut(cols, nbins=10, right_closed=True)
+--
+
+Cut all the columns in a Frame/f-expression by binning
+their values into equal width discrete intervals.
+
+Parameters
+----------
+cols: Frame | f-expression
+    Frame or f-expression consisting of numeric columns.
+nbins: int | list of ints | tuple of ints
+    When a single number is specified, this number of bins
+    will be used to bin each column of `cols`.
+    When a list or a tuple is provided, each column will be binned
+    by using its own number of bins. In the latter case,
+    the list/tuple length must be equal to the number of columns
+    in `cols`.
+right_closed: bool
+    Indicates whether bins should include the rightmost edge or not.
+
+Returns
+-------
+Frame/f-expression, where each column is filled with the respective bin ids.
+)"
+);
+
+
+/**
+  * Python-facing function that can take as an argument either a Frame or
+  * an f-expression.
+  */
+static oobj pyfn_cut(const PKArgs& args)
+{
+  if (args[0].is_none_or_undefined()) {
+    throw TypeError() << "Function `cut()` requires one positional argument, "
+                         "but none were given";
+  }
+  oobj arg0 = args[0].to_oobj();
+  oobj arg1 = args[1].is_none_or_undefined()? py::None() : args[1].to_oobj();
+  oobj arg2 = args[2].is_none_or_undefined()? py::True() : args[2].to_oobj();
+
+  if (arg0.is_frame()) {
+    return cut_frame(arg0, arg1, arg2);
```

Comment:
what if we had a generic method that takes `Expr(op, (DT,), params)` and converts it into `DT[:, Expr(op, (f[:],), params)]`.